File: clients/windows/src/ClipShare.Windows.Vault/VaultContractCodec.cs
namespace ClipShare.Windows.Vault;

using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.RegularExpressions;

public sealed record VaultEnvelopeContract
{
    public required int SchemaVersion { get; init; }

    public required string Algorithm { get; init; }

    public required string Purpose { get; init; }

    public required string VaultId { get; init; }

    public required string OriginDeviceId { get; init; }

    public required string EntityId { get; init; }

    public required string Field { get; init; }

    public required long KeyEpoch { get; init; }

    public required long NonceCounter { get; init; }

    public required string Nonce { get; init; }

    public required long PaddedPlaintextBytes { get; init; }

    public required string CipherAndTag { get; init; }
}

public static partial class VaultContractCodec
{
    private static readonly HashSet<string> Purposes =
    [
        "folder-metadata",
        "item-metadata",
        "item-payload",
        "file-key-wrap",
        "file-manifest",
        "file-chunk",
        "event-body",
        "epoch-wrapper",
    ];

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = false,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        UnmappedMemberHandling = JsonUnmappedMemberHandling.Disallow,
        AllowDuplicateProperties = false,
    };

    [GeneratedRegex("^[a-z][a-z0-9-]{0,31}$", RegexOptions.CultureInvariant)]
    private static partial Regex FieldPattern();

    [GeneratedRegex("^[A-Za-z0-9_-]+$", RegexOptions.CultureInvariant)]
    private static partial Regex Base64UrlPattern();

    public static VaultEnvelopeContract DecodeEnvelope(string json)
    {
        ArgumentNullException.ThrowIfNull(json);
        var envelope = JsonSerializer.Deserialize<VaultEnvelopeContract>(json, JsonOptions)
            ?? throw new JsonException("Vault envelope must be an object.");
        if (envelope.SchemaVersion != 1)
        {
            throw new InvalidDataException("Unsupported Vault envelope version.");
        }

        if (!string.Equals(envelope.Algorithm, "A256GCM", StringComparison.Ordinal) ||
            !Purposes.Contains(envelope.Purpose))
        {
            throw new InvalidDataException("Unsupported Vault encryption contract.");
        }

        _ = VaultId.Parse(envelope.VaultId);
        _ = DeviceId.Parse(envelope.OriginDeviceId);
        _ = VaultUuid.Validate(envelope.EntityId);
        if (!FieldPattern().IsMatch(envelope.Field) || envelope.KeyEpoch <= 0 || envelope.NonceCounter < 0)
        {
            throw new InvalidDataException("Invalid Vault envelope metadata.");
        }

        if (envelope.Purpose is not ("file-chunk" or "epoch-wrapper") && envelope.NonceCounter == 0)
        {
            throw new InvalidDataException("Record nonce counters start at one.");
        }

        byte[] encrypted = DecodeBase64Url(envelope.CipherAndTag);
        if (DecodeBase64Url(envelope.Nonce).Length != 12 || encrypted.Length < 16)
        {
            throw new InvalidDataException("Invalid Vault ciphertext shape.");
        }

        long plaintextBytes = encrypted.LongLength - 16;
        if (envelope.PaddedPlaintextBytes != plaintextBytes)
        {
            throw new InvalidDataException("Declared plaintext length does not match the ciphertext.");
        }

        if (envelope.Purpose != "file-chunk" &&
            (envelope.PaddedPlaintextBytes < 4096 || envelope.PaddedPlaintextBytes % 4096 != 0))
        {
            throw new InvalidDataException("Vault record plaintext must use 4 KiB buckets.");
        }

        if (envelope.Purpose == "file-chunk" && envelope.PaddedPlaintextBytes is < 0 or > 1_048_576)
        {
            throw new InvalidDataException("Invalid file chunk length.");
        }

        return envelope;
    }

    private static byte[] DecodeBase64Url(string value)
    {
        if (!Base64UrlPattern().IsMatch(value) || value.Length % 4 == 1)
        {
            throw new InvalidDataException("Invalid Base64URL.");
        }

        var padded = value.Replace('-', '+').Replace('_', '/').PadRight((value.Length + 3) / 4 * 4, '=');
        byte[] decoded;
        try
        {
            decoded = Convert.FromBase64String(padded);
        }
        catch (FormatException error)
        {
            throw new InvalidDataException("Invalid Base64URL.", error);
        }

        var canonical = Convert.ToBase64String(decoded).TrimEnd('=').Replace('+', '-').Replace('/', '_');
        if (!string.Equals(canonical, value, StringComparison.Ordinal))
        {
            throw new InvalidDataException("Non-canonical Base64URL.");
        }

        return decoded;
    }
}

File: clients/android/core/vault/src/main/kotlin/com/clipshare/core/vault/VaultContractCodec.kt
package com.clipshare.core.vault

import java.util.Base64
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

private val BASE64URL_PATTERN = Regex("^[A-Za-z0-9_-]+$")
private val FIELD_PATTERN = Regex("^[a-z][a-z0-9-]{0,31}$")
private val PURPOSES = setOf(
    "folder-metadata",
    "item-metadata",
    "item-payload",
    "file-key-wrap",
    "file-manifest",
    "file-chunk",
    "event-body",
    "epoch-wrapper",
)
private const val SCHEMA_VERSION = 1
private const val GCM_NONCE_BYTES = 12
private const val GCM_TAG_BYTES = 16
private const val MAXIMUM_FILE_CHUNK_BYTES = 1_048_576L
private const val PADDING_BUCKET_BYTES = 4096L
private const val BASE64_QUANTUM = 4

@Serializable
data class VaultEnvelopeContract(
    val schemaVersion: Int,
    val algorithm: String,
    val purpose: String,
    val vaultId: String,
    val originDeviceId: String,
    val entityId: String,
    val field: String,
    val keyEpoch: Long,
    val nonceCounter: Long,
    val nonce: String,
    val paddedPlaintextBytes: Long,
    val cipherAndTag: String,
)

object VaultContractCodec {
    private val json = Json {
        ignoreUnknownKeys = false
        isLenient = false
        explicitNulls = true
    }

    fun decodeEnvelope(value: String): VaultEnvelopeContract {
        rejectDuplicateTopLevelProperties(value)
        val envelope = json.decodeFromString<VaultEnvelopeContract>(value)
        require(envelope.schemaVersion == SCHEMA_VERSION) { "Unsupported Vault envelope version." }
        require(envelope.algorithm == "A256GCM") { "Unsupported Vault algorithm." }
        require(envelope.purpose in PURPOSES) { "Unknown Vault purpose." }
        VaultId.parse(envelope.vaultId)
        DeviceId.parse(envelope.originDeviceId)
        validateUuid(envelope.entityId)
        require(FIELD_PATTERN.matches(envelope.field)) { "Invalid field name." }
        require(envelope.keyEpoch > 0) { "Invalid key epoch." }
        require(envelope.nonceCounter >= 0) { "Invalid nonce counter." }
        if (envelope.purpose != "file-chunk" && envelope.purpose != "epoch-wrapper") {
            require(envelope.nonceCounter > 0) { "Record nonce counters start at one." }
        }
        require(decodeCanonicalBase64Url(envelope.nonce).size == GCM_NONCE_BYTES) { "Invalid nonce." }
        val encrypted = decodeCanonicalBase64Url(envelope.cipherAndTag)
        require(encrypted.size >= GCM_TAG_BYTES) { "Ciphertext must include a 16-byte tag." }
        val plaintextBytes = encrypted.size.toLong() - GCM_TAG_BYTES
        require(envelope.paddedPlaintextBytes == plaintextBytes) {
            "Declared plaintext length does not match the ciphertext."
        }
        if (envelope.purpose == "file-chunk") {
            require(envelope.paddedPlaintextBytes in 0..MAXIMUM_FILE_CHUNK_BYTES) { "Invalid chunk length." }
        } else {
            require(
                envelope.paddedPlaintextBytes >= PADDING_BUCKET_BYTES &&
                    envelope.paddedPlaintextBytes % PADDING_BUCKET_BYTES == 0L,
            ) { "Vault record plaintext must use 4 KiB buckets." }
        }
        return envelope
    }

    @Suppress("CyclomaticComplexMethod")
    private fun rejectDuplicateTopLevelProperties(value: String) {
        val seen = mutableSetOf<String>()
        var objectDepth = 0
        var arrayDepth = 0
        var stringStart = -1
        var escaped = false
        var propertyName = false
        var expectingProperty = false
        value.forEachIndexed { index, character ->
            if (stringStart >= 0) {
                when {
                    escaped -> escaped = false
                    character == '\\' -> escaped = true
                    character == '"' -> {
                        if (propertyName) {
                            val decoded = json.decodeFromString<String>(value.substring(stringStart, index + 1))
                            require(seen.add(decoded)) { "Duplicate Vault envelope property." }
                            expectingProperty = false
                        }
                        stringStart = -1
                    }
                }
            } else {
                when (character) {
                    '{' -> {
                        objectDepth += 1
                        if (objectDepth == 1 && arrayDepth == 0) expectingProperty = true
                    }
                    '}' -> objectDepth -= 1
                    '[' -> arrayDepth += 1
                    ']' -> arrayDepth -= 1
                    ',' -> if (objectDepth == 1 && arrayDepth == 0) expectingProperty = true
                    '"' -> {
                        stringStart = index
                        propertyName = objectDepth == 1 && arrayDepth == 0 && expectingProperty
                    }
                }
            }
        }
    }

    private fun decodeCanonicalBase64Url(value: String): ByteArray {
        require(BASE64URL_PATTERN.matches(value)) { "Invalid Base64URL." }
        require(value.length % BASE64_QUANTUM != 1) { "Invalid Base64URL length." }
        val decoded = Base64.getUrlDecoder().decode(value)
        require(Base64.getUrlEncoder().withoutPadding().encodeToString(decoded) == value) {
            "Non-canonical Base64URL."
        }
        return decoded
    }
}
